lib/router.js

```diff
@@ -0,0 +1,76 @@
+import { toChildArray, cloneElement } from 'preact'
+
+import { useEve } from './eve.js'
+import { frame } from './state.js'
+
+const { history, location } = window
+
+export const match = (route, path) => {
+  let props = {}
+  let key, val, i, j, v, k
+  key = val = ''
+  i = j = 0
+  while (i < route.length || j < path.length) {
+    v = path[i] || '/'
+    k = route[j] || '/'
+    if (k === '*') return (props[k] = path.slice(j)), props
+    if (val) {
+      if (v === '/' && k === '/') {
+        props[key] = val
+        val = ''
+        j++, i++
+      } else {
+        k !== '/' && ((key += k), j++)
+        v !== '/' && ((val += v), i++)
+      }
+    } else {
+      j++, i++
+      if (k === ':') (key = ''), (val = v)
+      else if (k !== v) return
+    }
+  }
+  return props
+}
+
+const shouldSkip = (event) =>
+  event.defaultPrevented ||
+  event.button ||
+  event.metaKey ||
+  event.altKey ||
+  event.ctrlKey ||
+  event.shiftKey
+
+export const RouterLink = (props) => {
+  props.onClick = (event) => {
+    if (shouldSkip(event)) return
+    event.preventDefault()
+    navigate(props.href)
+  }
+  return h('a', props)
+}
+
+const scores = {} // cache routes scores to avoid computations
+const byRank = (a, b) => rank(b.props.path || '') - rank(a.props.path || '')
+const rankPart = (acc, p) => acc + (p[0] === ':' ? 2 : (p !== '*') * 2 + 1)
+const rank = (route) =>
+  scores[route] ||
+  (scores[route] = route.split('/').filter(Boolean).reduce(rankPart, 0))
```

Comment:
this is the ranking system, we want to choose the path of the highest specificity, imagine those 3 routes:
- `*` a catch all for a routes to fall into
- `/admin/user/:userId/` edit a specific user info
- `/admin/user/new` create a new user

We obviously do not want all the routes to render the catch all, so we create a score based on parts of the route:

- `*` is worth 1
- a variable `:userId` here for example is worth 2
- a normal word is worth 3

this will give us the following scores:
- `*`  is `1`
- `/admin/user/:userId/` admin (3) + user (3) + `:userId` (2) so total of `8`
- `/admin/user/new` admin (3) + user (3) + new (3)

As such, the match will try to match the first the `/admin/user/new` route then `/admin/user/:userId/` and lastly the `*` catch all